src/LeafInputSignature.cpp

```diff
@@ -0,0 +1,403 @@
+#include "LeafInputSignature.hpp"
+#include "LeafInputSignatureOffline.hpp"
+#include "DagData.hpp"
+#include "RuntimeNode.hpp"
+#include "BuildQueue.hpp"
+#include "Profiler.hpp"
+#include "FileSign.hpp"
+#include "Scanner.hpp"
+#include "Driver.hpp"
+#include "Exec.hpp"
+#include "NodeResultPrinting.hpp"
+#include "AllBuiltNodes.hpp"
+#include "Atomic.hpp"
+#include <time.h>
+
+#if !TUNDRA_WIN32
+#include <unistd.h>
+#endif
+
+static bool FilterOutGeneratedIncludedFiles(void* _userData, const char* includingFile, const char* includedFile)
+{
+    //return true if we want to process the file, false if we want it to be skipped
+    return !IsFileGenerated((DagRuntimeData*)_userData, Djb2HashPath(includedFile), includedFile);
+}
+
+void CalculateLeafInputSignature(
+    BuildQueue* buildQueue,
+    const Frozen::DagNode* dagNode,
+    RuntimeNode* runtimeNode,
+    MemAllocLinear* scratch,
+    int profilerThreadId,
+    FILE* ingredient_stream)
+{
+    CheckDoesNotHaveLock(&buildQueue->m_Lock);
+
+    if (ingredient_stream)
+    {
+        time_t rawtime;
+        struct tm *info;
+        time( &rawtime );
+        info = localtime( &rawtime );
+
+        fprintf(ingredient_stream, "Current local time and date: %s", asctime(info));
+        fprintf(ingredient_stream, "Annotation %s\n\n", dagNode->m_Annotation.Get());
+        fprintf(ingredient_stream, "Cold hash ingredients (anything below this line is hashed and must not change to get a cache hit):\n");
+
+#if TUNDRA_WIN32
+        const char* username = getenv("USERNAME");
+        if (username != nullptr)
+            fprintf(ingredient_stream, "Signature generated by %s\n", username);
+#else
+        char hostname[1024];
```

Comment:
256 bytes is probably sufficient: https://web.archive.org/web/20190518124533/https://devblogs.microsoft.com/oldnewthing/?p=7873